File: internal/frc/tba_client.go
package frc

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const tbaBaseURL = "https://www.thebluealliance.com/api/v3"

// TBAClient wraps access to The Blue Alliance API.
type TBAClient struct {
	baseURL    string
	authKey    string
	httpClient *http.Client
}

// NewTBAClient creates a new TBA API client with the provided auth key.
func NewTBAClient(authKey string) *TBAClient {
	return &TBAClient{
		baseURL:    tbaBaseURL,
		authKey:    authKey,
		httpClient: &http.Client{Timeout: 20 * time.Second},
	}
}

func (c *TBAClient) getJSON(ctx context.Context, path string, query url.Values, out any) error {
	endpoint := c.baseURL + path
	if len(query) > 0 {
		endpoint += "?" + query.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return err
	}

	req.Header.Set("X-TBA-Auth-Key", c.authKey)
	req.Header.Set("Accept", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return fmt.Errorf("tba api %s returned %d: %s", path, resp.StatusCode, string(body))
	}

	dec := json.NewDecoder(resp.Body)
	return dec.Decode(out)
}

// OPRData contains OPR, DPR, and CCWM statistics for all teams at an event.
type OPRData struct {
	OPRs map[string]float64 `json:"oprs"`
	DPRs map[string]float64 `json:"dprs"`
	CCWMs map[string]float64 `json:"ccwms"`
}

// GetEventOPRs fetches OPR, DPR, and CCWM data for an event.
func (c *TBAClient) GetEventOPRs(ctx context.Context, eventKey string) (*OPRData, error) {
	var data OPRData
	path := fmt.Sprintf("/event/%s/oprs", eventKey)
	if err := c.getJSON(ctx, path, nil, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

// ComponentOPRData contains dynamic component OPR breakdown per team.
// TBA returns a map of component-name -> map[teamKey]value.
type ComponentOPRData struct {
	Components map[string]map[string]float64
}

func (d *ComponentOPRData) UnmarshalJSON(data []byte) error {
	var raw map[string]map[string]float64
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	d.Components = raw
	return nil
}

func (d *ComponentOPRData) componentMap(preferredNames []string, containsAll []string) map[string]float64 {
	if d == nil || len(d.Components) == 0 {
		return nil
	}

	lookup := make(map[string]map[string]float64, len(d.Components))
	for name, values := range d.Components {
		lookup[strings.ToLower(strings.TrimSpace(name))] = values
	}

	for _, name := range preferredNames {
		if values, ok := lookup[strings.ToLower(strings.TrimSpace(name))]; ok {
			return values
		}
	}

	// Fallback for new seasons: pick the largest component map that matches tokens.
	var best map[string]float64
	bestSize := -1
	for name, values := range lookup {
		match := true
		for _, token := range containsAll {
			if !strings.Contains(name, token) {
				match = false
				break
			}
		}
		if match && len(values) > bestSize {
			best = values
			bestSize = len(values)
		}
	}

	return best
}

// TeamPhaseOPRs returns best-effort auto/teleop/endgame point contributions for a team.
func (d *ComponentOPRData) TeamPhaseOPRs(teamKey string) (auto *float64, teleop *float64, endgame *float64) {
	autoMap := d.componentMap(
		[]string{"totalAutoPoints", "autoPoints", "Hub Auto Points"},
		[]string{"auto", "points"},
	)
	teleopMap := d.componentMap(
		[]string{"totalTeleopPoints", "teleopPoints", "Hub Teleop Points"},
		[]string{"teleop", "points"},
	)
	endgameMap := d.componentMap(
		[]string{"endGameTowerPoints", "endgamePoints", "Hub Endgame Points"},
		[]string{"endgame", "points"},
	)

	if autoMap != nil {
		if v, ok := autoMap[teamKey]; ok {
			auto = &v
		}
	}
	if teleopMap != nil {
		if v, ok := teleopMap[teamKey]; ok {
			teleop = &v
		}
	}
	if endgameMap != nil {
		if v, ok := endgameMap[teamKey]; ok {
			endgame = &v
		}
	}

	return auto, teleop, endgame
}

// GetEventComponentOPRs fetches component OPR breakdown for an event.
func (c *TBAClient) GetEventComponentOPRs(ctx context.Context, eventKey string) (*ComponentOPRData, error) {
	var data ComponentOPRData
	path := fmt.Sprintf("/event/%s/coprs", eventKey)
	if err := c.getJSON(ctx, path, nil, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

// RankingInfo contains ranking information for a team at an event.
type RankingInfo struct {
	TeamKey      string `json:"team_key"`
	Rank         int    `json:"rank"`
	MatchesPlayed int   `json:"matches_played"`
	QualAverage  *float64 `json:"qual_average"`
	ExtraStats   []float64 `json:"extra_stats"`
	SortOrders   []float64 `json:"sort_orders"`
	Record       struct {
		Wins   int `json:"wins"`
		Losses int `json:"losses"`
		Ties   int `json:"ties"`
	} `json:"record"`
	Dq         int `json:"dq"`
	QualPoints *int `json:"qual_points"`
	ElimPoints *int `json:"elim_points"`
	AwardPoints *int `json:"award_points"`
	AlliancePoints *int `json:"alliance_points"`
	TiePoints *int `json:"tie_points"`
	TotalPoints *int `json:"total_points"`
}

// EffectiveQualAverage returns qual_average or falls back to first sort order (year-specific ranking score).
func (r RankingInfo) EffectiveQualAverage() *float64 {
	if r.QualAverage != nil {
		return r.QualAverage
	}
	if len(r.SortOrders) > 0 {
		v := r.SortOrders[0]
		return &v
	}
	return nil
}

// EffectiveAvgMatchPoints returns average match points from sort_orders[1] if available.
func (r RankingInfo) EffectiveAvgMatchPoints() *float64 {
	if len(r.SortOrders) > 1 {
		v := r.SortOrders[1]
		return &v
	}
	return nil
}

// EffectiveTotalPoints returns total points, falling back to first extra stat (often total ranking points).
func (r RankingInfo) EffectiveTotalPoints() *int64 {
	if r.TotalPoints != nil {
		v := int64(*r.TotalPoints)
		return &v
	}
	if len(r.ExtraStats) > 0 {
		v := int64(math.Round(r.ExtraStats[0]))
		return &v
	}
	return nil
}

// EffectiveQualPoints returns qual points or falls back to effective total points when only extra_stats are provided.
func (r RankingInfo) EffectiveQualPoints() *int64 {
	if r.QualPoints != nil {
		v := int64(*r.QualPoints)
		return &v
	}
	return r.EffectiveTotalPoints()
}

func optionalIntToInt64Ptr(v *int) *int64 {
	if v == nil {
		return nil
	}
	out := int64(*v)
	return &out
}

func (r RankingInfo) EffectiveElimPoints() *int64 {
	return optionalIntToInt64Ptr(r.ElimPoints)
}

func (r RankingInfo) EffectiveAwardPoints() *int64 {
	return optionalIntToInt64Ptr(r.AwardPoints)
}

func (r RankingInfo) EffectiveAlliancePoints() *int64 {
	return optionalIntToInt64Ptr(r.AlliancePoints)
}

// EventRankings fetches rankings for an event.
func (c *TBAClient) GetEventRankings(ctx context.Context, eventKey string) ([]RankingInfo, error) {
	var data struct {
		Rankings []RankingInfo `json:"rankings"`
	}
	path := fmt.Sprintf("/event/%s/rankings", eventKey)
	if err := c.getJSON(ctx, path, nil, &data); err != nil {
		return nil, err
	}
	return data.Rankings, nil
}

// EventInfo contains event metadata.
type EventInfo struct {
	Key       string     `json:"key"`
	Name      string     `json:"name"`
	EventCode string     `json:"event_code"`
	Year      int        `json:"year"`
	StartDate string     `json:"start_date"`
	EndDate   string     `json:"end_date"`
	Timezone  string     `json:"timezone"`
	Official  bool       `json:"official"`
	Playoff   string     `json:"playoff"`
}

// GetEvent fetches detailed event information.
func (c *TBAClient) GetEvent(ctx context.Context, eventKey string) (*EventInfo, error) {
	var event EventInfo
	path := fmt.Sprintf("/event/%s", eventKey)
	if err := c.getJSON(ctx, path, nil, &event); err != nil {
		return nil, err
	}
	return &event, nil
}

// MatchInfo contains match information including timing.
type MatchInfo struct {
	Key          string    `json:"key"`
	EventKey     string    `json:"event_key"`
	CompLevel    string    `json:"comp_level"`
	SetNumber    int       `json:"set_number"`
	MatchNumber  int       `json:"match_number"`
	Alliances    struct {
		Red struct {
			Teams []string `json:"team_keys"`
			Score int      `json:"score"`
		} `json:"red"`
		Blue struct {
			Teams []string `json:"team_keys"`
			Score int      `json:"score"`
		} `json:"blue"`
	} `json:"alliances"`
	ActualTime int64  `json:"actual_time"`
	PredictedTime int64  `json:"predicted_time"`
	ScheduledTime int64  `json:"scheduled_time"`
	ScoreBreakdown interface{} `json:"score_breakdown"`
}

// GetEventMatches fetches all matches for an event.
func (c *TBAClient) GetEventMatches(ctx context.Context, eventKey string) ([]MatchInfo, error) {
	var matches []MatchInfo
	path := fmt.Sprintf("/event/%s/matches", eventKey)
	if err := c.getJSON(ctx, path, nil, &matches); err != nil {
		return nil, err
	}
	return matches, nil
}

File: dotnet/TealTeam.Web/Services/TbaClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TealTeam.Web.Services;

// DTOs and client for The Blue Alliance API v3 (port of internal/frc/tba_client.go)

public class OprData
{
    [JsonPropertyName("oprs")] public Dictionary<string, double> Oprs { get; set; } = new();
    [JsonPropertyName("dprs")] public Dictionary<string, double> Dprs { get; set; } = new();
    [JsonPropertyName("ccwms")] public Dictionary<string, double> Ccwms { get; set; } = new();
}

/// <summary>
/// Dynamic component OPR breakdown: component-name -> teamKey -> value.
/// Includes fallback matching for season schema variance.
/// </summary>
public class ComponentOprData
{
    public Dictionary<string, Dictionary<string, double>> Components { get; set; } = new();

    private Dictionary<string, double>? ComponentMap(string[] preferredNames, string[] containsAll)
    {
        if (Components.Count == 0) return null;

        var lookup = Components.ToDictionary(
            kv => kv.Key.Trim().ToLowerInvariant(), kv => kv.Value);

        foreach (var name in preferredNames)
        {
            if (lookup.TryGetValue(name.Trim().ToLowerInvariant(), out var values))
            {
                return values;
            }
        }

        // Fallback for new seasons: pick the largest component map matching all tokens.
        Dictionary<string, double>? best = null;
        var bestSize = -1;
        foreach (var (name, values) in lookup)
        {
            if (containsAll.All(name.Contains) && values.Count > bestSize)
            {
                best = values;
                bestSize = values.Count;
            }
        }

        return best;
    }

    public (double? Auto, double? Teleop, double? Endgame) TeamPhaseOprs(string teamKey)
    {
        var autoMap = ComponentMap(
            new[] { "totalAutoPoints", "autoPoints", "Hub Auto Points" },
            new[] { "auto", "points" });
        var teleopMap = ComponentMap(
            new[] { "totalTeleopPoints", "teleopPoints", "Hub Teleop Points" },
            new[] { "teleop", "points" });
        var endgameMap = ComponentMap(
            new[] { "endGameTowerPoints", "endgamePoints", "Hub Endgame Points" },
            new[] { "endgame", "points" });

        double? auto = autoMap != null && autoMap.TryGetValue(teamKey, out var a) ? a : null;
        double? teleop = teleopMap != null && teleopMap.TryGetValue(teamKey, out var t) ? t : null;
        double? endgame = endgameMap != null && endgameMap.TryGetValue(teamKey, out var e) ? e : null;
        return (auto, teleop, endgame);
    }
}

public class RankingRecord
{
    [JsonPropertyName("wins")] public int Wins { get; set; }
    [JsonPropertyName("losses")] public int Losses { get; set; }
    [JsonPropertyName("ties")] public int Ties { get; set; }
}

public class RankingInfo
{
    [JsonPropertyName("team_key")] public string TeamKey { get; set; } = "";
    [JsonPropertyName("rank")] public int Rank { get; set; }
    [JsonPropertyName("matches_played")] public int MatchesPlayed { get; set; }
    [JsonPropertyName("qual_average")] public double? QualAverage { get; set; }
    [JsonPropertyName("extra_stats")] public List<double> ExtraStats { get; set; } = new();
    [JsonPropertyName("sort_orders")] public List<double> SortOrders { get; set; } = new();
    [JsonPropertyName("record")] public RankingRecord Record { get; set; } = new();
    [JsonPropertyName("dq")] public int Dq { get; set; }
    [JsonPropertyName("qual_points")] public int? QualPoints { get; set; }
    [JsonPropertyName("elim_points")] public int? ElimPoints { get; set; }
    [JsonPropertyName("award_points")] public int? AwardPoints { get; set; }
    [JsonPropertyName("alliance_points")] public int? AlliancePoints { get; set; }
    [JsonPropertyName("tie_points")] public int? TiePoints { get; set; }
    [JsonPropertyName("total_points")] public int? TotalPoints { get; set; }

    // Fallback extraction for year-specific ranking schema variance.
    public double? EffectiveQualAverage()
        => QualAverage ?? (SortOrders.Count > 0 ? SortOrders[0] : null);

    public double? EffectiveAvgMatchPoints()
        => SortOrders.Count > 1 ? SortOrders[1] : null;

    public long? EffectiveTotalPoints()
    {
        if (TotalPoints != null) return TotalPoints.Value;
        if (ExtraStats.Count > 0) return (long)Math.Round(ExtraStats[0]);
        return null;
    }

    public long? EffectiveQualPoints()
        => QualPoints != null ? QualPoints.Value : EffectiveTotalPoints();

    public long? EffectiveElimPoints() => ElimPoints;
    public long? EffectiveAwardPoints() => AwardPoints;
    public long? EffectiveAlliancePoints() => AlliancePoints;
}

public class TbaAlliance
{
    [JsonPropertyName("team_keys")] public List<string> Teams { get; set; } = new();
    [JsonPropertyName("score")] public int Score { get; set; }
}

public class TbaAlliances
{
    [JsonPropertyName("red")] public TbaAlliance Red { get; set; } = new();
    [JsonPropertyName("blue")] public TbaAlliance Blue { get; set; } = new();
}

public class MatchInfo
{
    [JsonPropertyName("key")] public string Key { get; set; } = "";
    [JsonPropertyName("event_key")] public string EventKey { get; set; } = "";
    [JsonPropertyName("comp_level")] public string CompLevel { get; set; } = "";
    [JsonPropertyName("set_number")] public int SetNumber { get; set; }
    [JsonPropertyName("match_number")] public int MatchNumber { get; set; }
    [JsonPropertyName("alliances")] public TbaAlliances Alliances { get; set; } = new();
    [JsonPropertyName("actual_time")] public long ActualTime { get; set; }
    [JsonPropertyName("predicted_time")] public long PredictedTime { get; set; }
    [JsonPropertyName("scheduled_time")] public long ScheduledTime { get; set; }
    [JsonPropertyName("score_breakdown")] public JsonElement? ScoreBreakdown { get; set; }
}

public class TbaClient
{
    private const string BaseUrl = "https://www.thebluealliance.com/api/v3";

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(20) };
    private static readonly JsonSerializerOptions JsonOptions = new() { PropertyNameCaseInsensitive = true };

    private readonly string _authKey;

    public TbaClient(string authKey)
    {
        _authKey = authKey;
    }

    private async Task<T> GetJsonAsync<T>(string path, CancellationToken ct)
    {
        if (!Connectivity.ShouldSkipConnectivityCheck(BaseUrl))
        {
            try
            {
                await Connectivity.EnsureInternetForBaseUrlAsync(BaseUrl, ct);
            }
            catch (InternetUnavailableException ex)
            {
                Connectivity.RecordApiError(ex);
                throw;
            }
        }

        var endpoint = BaseUrl + path;
        Exception? lastError = null;
        for (var attempt = 0; attempt < Connectivity.ApiRetryMaxAttempts; attempt++)
        {
            try
            {
                using var req = new HttpRequestMessage(HttpMethod.Get, endpoint);
                req.Headers.TryAddWithoutValidation("X-TBA-Auth-Key", _authKey);
                req.Headers.TryAddWithoutValidation("Accept", "application/json");

                using var resp = await Http.SendAsync(req, ct);
                if (!resp.IsSuccessStatusCode)
                {
                    var body = await resp.Content.ReadAsStringAsync(ct);
                    var err = new HttpRequestException(
                        $"tba api {path} returned {(int)resp.StatusCode}: {body[..Math.Min(body.Length, 4096)]}");
                    Connectivity.RecordApiError(err);
                    lastError = err;
                    if (attempt < Connectivity.ApiRetryMaxAttempts - 1 &&
                        Connectivity.ShouldRetryStatusCode((int)resp.StatusCode))
                    {
                        await Task.Delay(Connectivity.BackoffDelay(attempt), ct);
                        continue;
                    }
                    throw err;
                }

                await using var stream = await resp.Content.ReadAsStreamAsync(ct);
                var result = await JsonSerializer.DeserializeAsync<T>(stream, JsonOptions, ct)
                    ?? throw new JsonException($"tba api {path} returned empty payload");
                Connectivity.RecordApiSuccess();
                return result;
            }
            catch (HttpRequestException ex) when (attempt < Connectivity.ApiRetryMaxAttempts - 1 && ex.StatusCode == null)
            {
                Connectivity.RecordApiError(ex);
                lastError = ex;
                await Task.Delay(Connectivity.BackoffDelay(attempt), ct);
            }
        }

        var finalErr = lastError ?? new HttpRequestException($"tba api {path} retries exhausted");
        Connectivity.RecordApiError(finalErr);
        throw finalErr;
    }

    public Task<OprData> GetEventOprsAsync(string eventKey, CancellationToken ct = default)
        => GetJsonAsync<OprData>($"/event/{eventKey}/oprs", ct);

    public async Task<ComponentOprData?> GetEventComponentOprsAsync(string eventKey, CancellationToken ct = default)
    {
        var raw = await GetJsonAsync<Dictionary<string, Dictionary<string, double>>>($"/event/{eventKey}/coprs", ct);
        return new ComponentOprData { Components = raw };
    }

    private class RankingsResponse
    {
        [JsonPropertyName("rankings")] public List<RankingInfo> Rankings { get; set; } = new();
    }

    public async Task<List<RankingInfo>> GetEventRankingsAsync(string eventKey, CancellationToken ct = default)
        => (await GetJsonAsync<RankingsResponse>($"/event/{eventKey}/rankings", ct)).Rankings;

    public Task<List<MatchInfo>> GetEventMatchesAsync(string eventKey, CancellationToken ct = default)
        => GetJsonAsync<List<MatchInfo>>($"/event/{eventKey}/matches", ct);
}
